### ISSUE TITLE
GLFW Backend and it’s place within the library

### ISSUE THREAD

**Issue by eliasdaler** [AllenDang/cimgui-go#41]
Currently, we have a GLFW backend as a part of "main" repo, but should it stay that way?

I think it might be better to move it to "examples" directory, because you *can* use cimgui-go without it by implementing another backend with a lib of your choice. 

For example, I managed to get cimgui-go working with Ebitengine. I even believe it might be a "better" example, because it’s (almost) a pure Go library, plays better with Go ecosystem, and doesn’t require you to use CGo (so the only CGo code left is between your program and Dear ImGui itself). 

So, my proposal is this:

A. Move glfw backend to "example/glfw" directory. 

OR 

B. If you want to make glfw backend a reusable package, I believe that making it a separate repo might be better (kinda like imgui-go did with [this repo](https://github.com/inkyblackness/imgui-go-examples, but don’t make backend impl "internal")

**Comment by AllenDang** [comment]
@eliasdaler The reason why I start all this is, neither imgui nor imgui-go provides a ready-to-use library, they all need me to find or implement a backend to get everything work.
This repo and giu should provide a gui framework which could be used immediately without extra work.
That's why I added a default backend implementation inside with a build flag to remove it.

**Comment by eliasdaler** [comment]
Can't the backend be moved to a separate repo/package or inside giu itself?
cimgui-go should be usable outside of other contents, and I'd prefer it I didn't have to do anything to remove the things I don't need (like implot which I don't need... but that's another issue.).

If the user wants to have a GLFW backend, they can import "glfwbackend" package. It the user wants pure cimgui-go binding, they can import "cimgui". 

Ideally, the output of cimgui-go should be treated as a simple Go package which can function without any other things. And currently, Go module system doesn't allow you to "remove" features it they're a part of the package. It only allows you to separate one big package into several, which is what I'd prefer.

**Comment by gucio321** [comment]
In my opinion the perfect scenario would be that cimgui-go should have a backend implementation set by default, for people who wants just "plug-and-play" this package. but it should be a method "SetBackend(backend Backend)" that allows to change the default implementation.
This way, we just need an interface `Backend` that exposes the necessary methods (just like the old AllenDang/imgui-go had)

**Comment by eliasdaler** [comment]
I think that cimgui-go might have the following structure:

```
./
    glfwbackend/
        backend.go
        // and other files of backend impl
    funcs.go
    types.go // and the rest of "wrapper" stuff
```

This way, if people want to use the backend, they can do this:

```go
import "github.com/AllenDang/cimgui-go/glfwbackend"

...

glfwbackend.Start()
```

It is "plug-and-play", but it makes cimgui-go not grab GLFW for me if I don't ever need it. I only need cimgui bindings.

**Comment by gucio321** [comment]
or even better:
```txt
.
└── pkg
    ├── backend
    │   └── glfw
    └── cimgui
```

This way, go will not grab (iirc) glfw if it is not imported, it will just use cimgui

**Comment by eliasdaler** [comment]
I don't like "pkg" directory too much. Most good packages don't require you to import them like this:

```go
import "github.com/AllenDang/cimgui-go/pkg/cimgui"
```

You just import them like this:

```go
import "github.com/AllenDang/cimgui-go"
```

And I think that it would still import GLFW unless "backend" package has a separate go.mod, so at this point it's better to move backend implementation to a separate repo or make a separate "go.mod" for it.

**Comment by eliasdaler** [comment]
@AllenDang now that we've decided to remove GLFW backend, how should I do it?
I think cimgui-go might contain ebiten-imgui inside of it until its ported to cimgui-go in upstream (should be quick), but what do we do after?

Do we leave the example/main.go which shows how to use cimgui-go and ebiten-imgui? I don't think we need to store the backend itself inside this repo, it will be redundant.

**Comment by AllenDang** [comment]
@eliasdaler I prefer to store the backend into this repo and use a build flag to remove it, like what it is for GLFW backend.

**Comment by eliasdaler** [comment]
I wouldn't be comfortable by taking someone's code and putting it into another repo essentially making a hard fork of it.
Guess I'll try to rewrite a GLFW backend to live in another package so that users who `go get` cimgui-go don't have to depend on GLFW or this backend...

**Comment by AllenDang** [comment]
So you suggest we create a new repo say it's named "cimgui-go-backend-ebiten", and add a example code inside cimgui-go to demonstrate the usage, and add a section to readme  like "backend"?

**Comment by eliasdaler** [comment]
No, I suggest to add "examples" directory here and demonstrate the usage there.
Importing "ebiten-imgui" would be enough to make it work so we don't need "cimgui-go-backend-ebiten" to not duplicate code of "ebiten-imgui" repo. 

I would be okay with making "cimgui-go-backend-glfw" repo, though.

**Comment by AllenDang** [comment]
I think we will eventually need to modify the backend, let me create the repo now.

**Comment by eliasdaler** [comment]
If we need to modify it, we can just send PRs to ebiten-imgui repo...

**Comment by gucio321** [comment]
yah, and (if something unexpected happens) can fork the upstream and use it with `go mod edit -replace`

**Comment by AllenDang** [comment]
ebiten backend provides a chance to solve two major issues which comes with imgui's GLFW backend:
1. LoadTexture in a coroutine. Since imgui's GLFW backend launch opengl in main thread, it is very hard to implement loading texture in a coroutine.
2. A proper video player.

**Comment by Chillance** [comment]
So, I'm getting this now currently:
```
backend.cpp:9:10: fatal error: thirdparty/glfw/include/GLFW/glfw3.h: No such file or directory
    9 | #include "thirdparty/glfw/include/GLFW/glfw3.h" // Will drag system OpenGL headers
```

I'm guessing it's because it's a submodule and not not included when building this as a library.

**Comment by gucio321** [comment]
oh no, so we need to revert the other PR as well?!

**Comment by eliasdaler** [comment]
@gucio321 yes

@Chillance note that GLFW backend will soon be replaced in favor of ebiten-imgui.

**Comment by gucio321** [comment]
> @gucio321 yes

well... let's go ahead and do it

**Comment by GoWeasel** [comment]
Interesting discussion here

**Who is the lib user?** on long term i would suggest, **both**, 
- (A) the one who just need to create a quick simple tool for easy tasks,
- (B) the one who builds up a bigger project.

**Backend flexibility ?**
- should be a must have, like in imgui
- of course for the (A) User just take 1 default backend, and all other optional like custom backend.

**About Ebiten:** 
nice for 2D thinks, but i need 3D on div parts, so not my choice,

just replacing it as Eliasdaler announced to Chillance is a really bad idea, 
better do such thing on a seperate repro, and make cimgui-go flexibel to use different backends

at today nobody knows which backends will fit as best in feature, so a flex base would be great,
as example: think about perhaps future easier vulkan ways?

**Comment by GoWeasel** [comment]
>ebiten backend provides a chance to solve two major issues which comes with imgui's GLFW backend:

>LoadTexture in a coroutine. Since imgui's GLFW backend launch opengl in main thread, it is very hard to implement loading texture in a coroutine.

is that not more a general opengl single thread problem than a glfw problem?, prob. take vulkan

![image](https://user-images.githubusercontent.com/61587283/214123094-371dfef9-7408-4c39-9be7-cf884ba08918.png)

**Comment by gucio321** [comment]
> Who is the lib user? on long term i would suggest, both,
>
> (A) the one who just need to create a quick simple tool for easy tasks,
> (B) the one who builds up a bigger project.

Don't forget about [giu](https://github.com/AllenDang/giu)
In my opinion this framework should be for these users, who REALLY need greate customization abilities.
These one, who just want to create a simple UI should use giu rather than cimgui-go

> Backend flexibility ?
>
> should be a must have, like in imgui
> of course for the (A) User just take 1 default backend, and all other optional like custom backend.

technically, we should have an interface with all the metods backend needs to expose.
Over that, the default backend is (just like you said) "must have". I suppose that most users would like to have such a basic thing already implemented so that they don't need to play 20 hours to implement them.

**Comment by gucio321** [comment]
also, I'm just looking at this and I'm not sure if cimgui-go actually uses backend somehow. As far as I see it is only used by examples and is mostly encapsulated in `backend.go`
IMO in order to solve this somehow, the first thing we're supposed to do is extract current backend implementation into a separated package

**Comment by GoWeasel** [comment]
> Don't forget about [giu](https://github.com/AllenDang/giu)
> In my opinion this framework should be for these users, who REALLY need greate customization abilities.
> These one, who just want to create a simple UI should use giu rather than cimgui-go

The biggest problem on giu is the different syntax, when u see an example code off imgui code, than u first need to refactor to giu code (latest on this point, you wished directly to started with cimgui-go instead of giu), concerning this point, the project is just an archive in my eyes.

Therefore the new way (cimgui-go) with more compareable syntax to imgui, with less differences, is much **better.**
Of course there are still some points, concerning languages diffs, but mostly u can just write down and copy parts (right window = c++ imgui / left window = golang imgui project)  :-)

**Comment by GoWeasel** [comment]
> also, I'm just looking at this and I'm not sure if cimgui-go actually uses backend somehow. As far as I see it is only used by examples and is mostly encapsulated in `backend.go`
> IMO in order to solve this somehow, the first thing we're supposed to do is extract current backend implementation into a separated package

as more interoperability or flexibilty (backend choice etc) is build in cimgui-go (as long term), as more users could use it, for many different usecases, take a look at the matrix messenger project as example. 

one messenger -> Whatsapp, Signal, Threma, Slack, etc..... new one... okay just code a new bridge .... fine works
of course it can happens, that u then perhaps did not have 100% feature of the single app, but when u reach 80% that is more worth than to have 5 apps for messaging. (greetings from pareto principle) :yum:

**Comment by eliasdaler** [comment]
Backend is encapsulated, but it would be better if it lived in another repo so that cimgui-go remains compact and users who don't use GLFW won't need to clone glfw dir.

**Comment by gucio321** [comment]
@AllenDang we should deffinitely switch to go-gl/glfw because:
- we shouldn't focus on binding - there is a dedicated team for this
- if we have some idea how to improve glfw wrapper why can't we just go and open an issue there
- this will not hit performence in general (in my opinion) because: now we recompile whole glfw (well theoritically we compile only our C binding) and when using go-gl/glfw we should compile it once and then we don't care about it (as it is cached in go cache)

**Comment by AllenDang** [comment]
@gucio321 I just recalled a bad memory back days, when Imgui released a new version, I need to go through the change logs line by line to check whether any backend stuff is changed, sometime I might just miss a single sentence about that, the backend implemented via go-gl/glfw will be broken.

As Imgui is keep evolving, these things could happen time to time.

That's why I intend to reuse the backend and extend from it in cimgui-go, as the description says, this should be a auto-generated wrapper, if we need to implement a custom backend, I prefer to do it in giu, you think?

**Comment by gucio321** [comment]
Yes and no. Currently we our custom glfw wrapper. Can't we just use this one from go-gl? An use imgui_impl_glfw to go imgui things... (Ifk if its gonna eork, what do u think?)
Problem with current implementation is that it is bugged as hell and generally if there is any chance to stitch to qualified version of this wrapper I'd definitely recommend to use it

**Comment by gucio321** [comment]
@AllenDang We can take this C++ code and try to write equivalent using go-gl/glfw and wrapped imgui funcs, can't we?
<details>

```c++
GLFWwindow *igCreateGLFWWindow(const char *title, int width, int height, GLFWWindowFlags flags,
                               VoidCallback afterCreateContext) {
  // Setup window
  glfwSetErrorCallback(glfw_error_callback);
  if (!glfwInit())
    return NULL;

    // Decide GL+GLSL versions
#if defined(IMGUI_IMPL_OPENGL_ES2)
  // GL ES 2.0 + GLSL 100
  const char *glsl_version = "#version 100";
  glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 2);
  glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0);
  glfwWindowHint(GLFW_CLIENT_API, GLFW_OPENGL_ES_API);
#elif defined(__APPLE__)
  // GL 3.2 + GLSL 150
  const char *glsl_version = "#version 150";
  glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
  glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 2);
  glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE); // 3.2+ only
  glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);           // Required on Mac
#else
  // GL 3.0 + GLSL 130
  const char *glsl_version = "#version 130";
  glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
  glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0);
  // glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);  // 3.2+
  // only glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE); // 3.0+ only
#endif

  if ((flags & GLFWWindowFlagsNotResizable) != 0) {
    glfwWindowHint(GLFW_RESIZABLE, GLFW_FALSE);
  }

  if ((flags & GLFWWindowFlagsMaximized) != 0) {
    glfwWindowHint(GLFW_MAXIMIZED, GLFW_TRUE);
  }

  if ((flags & GLFWWindowFlagsFloating) != 0) {
    glfwWindowHint(GLFW_FLOATING, GLFW_TRUE);
  }

  if ((flags & GLFWWindowFlagsFrameless) != 0) {
    glfwWindowHint(GLFW_FLOATING, GLFW_TRUE);
  }

  if ((flags & GLFWWindowFlagsTransparent) != 0) {
    glfwWindowHint(GLFW_TRANSPARENT_FRAMEBUFFER, GLFW_TRUE);
  }

  // Create window with graphics context
  GLFWwindow *window = glfwCreateWindow(width, height, title, NULL, NULL);
  if (window == NULL)
    return NULL;
  glfwMakeContextCurrent(window);
  glfwSwapInterval(1); // Enable vsync

  // Setup Dear ImGui context
  igCreateContext(0);

  if (afterCreateContext != NULL) {
    afterCreateContext();
  }

  ImGuiIO *io = igGetIO();
  io->ConfigFlags |= ImGuiConfigFlags_NavEnableKeyboard; // Enable Keyboard Controls
  // io.ConfigFlags |= ImGuiConfigFlags_NavEnableGamepad;      // Enable Gamepad
  // Controls
  io->ConfigFlags |= ImGuiConfigFlags_DockingEnable;   // Enable Docking
  io->ConfigFlags |= ImGuiConfigFlags_ViewportsEnable; // Enable Multi-Viewport
                                                       // / Platform Windows
  // io.ConfigViewportsNoAutoMerge = true;
  // io.ConfigViewportsNoTaskBarIcon = true;

  io->IniFilename = "";

  // Setup Dear ImGui style
  igStyleColorsDark(0);
  // igStyleColorsLight();

  // When viewports are enabled we tweak WindowRounding/WindowBg so platform
  // windows can look identical to regular ones.
  ImGuiStyle *style = igGetStyle();
  if (io->ConfigFlags & ImGuiConfigFlags_ViewportsEnable) {
    style->WindowRounding = 0.0f;
    style->Colors[ImGuiCol_WindowBg].w = 1.0f;
  }

  // Setup Platform/Renderer backends
  ImGui_ImplGlfw_InitForOpenGL(window, true);
  ImGui_ImplOpenGL3_Init(glsl_version);

  // Install extra callback
  glfwSetWindowRefreshCallback(window, glfw_window_refresh_callback);
  glfwMakeContextCurrent(NULL);
  return window;
}

void glfw_render(GLFWwindow *window, VoidCallback renderLoop) {
  // Start the Dear ImGui frame
  ImGui_ImplOpenGL3_NewFrame();
  ImGui_ImplGlfw_NewFrame();
  igNewFrame();

  glfwSetWindowUserPointer(window, (void *)renderLoop);

  // Do ui stuff here
  if (renderLoop != NULL) {
    renderLoop();
  }

  // Rendering
  igRender();
  int display_w, display_h;
  glfwGetFramebufferSize(window, &display_w, &display_h);
  glViewport(0, 0, display_w, display_h);
  glClearColor(clear_color.x * clear_color.w, clear_color.y * clear_color.w, clear_color.z * clear_color.w,
               clear_color.w);
  glClear(GL_COLOR_BUFFER_BIT);
  ImGui_ImplOpenGL3_RenderDrawData(igGetDrawData());

  ImGuiIO *io = igGetIO();

  // Update and Render additional Platform Windows
  // (Platform functions may change the current OpenGL context, so we
  // save/restore it to make it easier to paste this code elsewhere.
  //  For this specific demo app we could also call
  //  glfwMakeContextCurrent(window) directly)
  if (io->ConfigFlags & ImGuiConfigFlags_ViewportsEnable) {
    GLFWwindow *backup_current_context = glfwGetCurrentContext();
    igUpdatePlatformWindows();
    igRenderPlatformWindowsDefault(NULL, NULL);
    glfwMakeContextCurrent(backup_current_context);
  }

  glfwSwapBuffers(window);
}

void igRunLoop(GLFWwindow *window, VoidCallback loop, VoidCallback beforeRender, VoidCallback afterRender,
               VoidCallback beforeDestroyContext) {
  glfwMakeContextCurrent(window);
  ImGuiIO *io = igGetIO();

  // Load Fonts
  // - If no fonts are loaded, dear imgui will use the default font. You can
  // also load multiple fonts and use igPushFont()/PopFont() to select them.
  // - AddFontFromFileTTF() will return the ImFont* so you can store it if you
  // need to select the font among multiple.
  // - If the file cannot be loaded, the function will return NULL. Please
  // handle those errors in your application (e.g. use an assertion, or display
  // an error and quit).
  // - The fonts will be rasterized at a given size (w/ oversampling) and stored
  // into a texture when calling ImFontAtlas::Build()/GetTexDataAsXXXX(), which
  // ImGui_ImplXXXX_NewFrame below will call.
  // - Read 'docs/FONTS.md' for more instructions and details.
  // - Remember that in C/C++ if you want to include a backslash \ in a string
  // literal you need to write a double backslash \\ !
  // io.Fonts->AddFontDefault();
  // io.Fonts->AddFontFromFileTTF("../../misc/fonts/Roboto-Medium.ttf", 16.0f);
  // io.Fonts->AddFontFromFileTTF("../../misc/fonts/Cousine-Regular.ttf", 15.0f);
  // io.Fonts->AddFontFromFileTTF("../../misc/fonts/DroidSans.ttf", 16.0f);
  // io.Fonts->AddFontFromFileTTF("../../misc/fonts/ProggyTiny.ttf", 10.0f);
  // ImFont* font =
  // io.Fonts->AddFontFromFileTTF("c:\\Windows\\Fonts\\ArialUni.ttf", 18.0f,
  // NULL, io.Fonts->GetGlyphRangesJapanese()); IM_ASSERT(font != NULL);

  // Main loop
  double lasttime = glfwGetTime();
  while (!glfwWindowShouldClose(window)) {
    if (beforeRender != NULL) {
      beforeRender();
    }

    glfw_render(window, loop);

    while (glfwGetTime() < lasttime + 1.0 / glfw_target_fps) {
      // do nothing here
    }
    lasttime += 1.0 / glfw_target_fps;

    if (extra_frame_count > 0) {
      extra_frame_count--;
    } else {
      glfwWaitEvents();
      extra_frame_count = MAX_EXTRA_FRAME_COUNT;
    }

    glfwPollEvents();

    if (afterRender != NULL) {
      afterRender();
    }
  }

  // Cleanup
  ImGui_ImplOpenGL3_Shutdown();
  ImGui_ImplGlfw_Shutdown();

  if (beforeDestroyContext != NULL) {
    beforeDestroyContext();
  }

  igDestroyContext(0);

  glfwDestroyWindow(window);
  glfwTerminate();

  return;
}
```

</details>

**Comment by gucio321** [comment]
@AllenDang the longer I look at it, the more am I convinced of this refactor being possible.
I did a dev branch on https://github.com/gucio321/cimgui-go/tree/backend-refactor
It fails builds yet, but should be possible to fix...

**Comment by AllenDang** [comment]
> @AllenDang We can take this C++ code and try to write equivalent using go-gl/glfw and wrapped imgui funcs, can't we?

That's what I did in giu, translate line by line.

**Comment by gucio321** [comment]
meh, I think I understand.
So in current impl, if something doesn't work you can just copy code from imgui example right?

**Comment by eliasdaler** [comment]
Why is glfw needed at all? Is it for "example" + "backend" part?
The Dear ImGui wrapper should be separate from backends, since you might want to use cimgui with Ebitengine and won't need to touch glfw ever.

**Comment by eliasdaler** [comment]
I feel like this discussion got sidetracked. My issue was only the placement of glfw backend which should be in a separate package and not required for import for everyone, since you can probably use cimgui-go with different backends.

**Comment by gucio321** [comment]
@AllenDang so could you please try to address #209 and #194

**Comment by damntourists** [comment]
Thought I should mention that when I removed glfw, it still produced these ld errors:
```
/home/brett/projects/sdk/go1.22.0/pkg/tool/linux_amd64/link: running g++ failed: exit status 1
/usr/bin/ld: /tmp/go-link-161865402/000014.o: in function `_cgo_cdf43a2df1d5_Cfunc_igCreateTexture':
/tmp/go-build/sdl_backend.cgo2.c:77:(.text+0x5d): undefined reference to `igCreateTexture'
/usr/bin/ld: /tmp/go-link-161865402/000014.o: in function `_cgo_cdf43a2df1d5_Cfunc_igDeleteTexture':
/tmp/go-build/sdl_backend.cgo2.c:92:(.text+0x88): undefined reference to `igDeleteTexture'
/usr/bin/ld: /tmp/go-link-161865402/000014.o: in function `_cgo_cdf43a2df1d5_Cfunc_igRefresh':
/tmp/go-build/sdl_backend.cgo2.c:122:(.text+0xc7): undefined reference to `igRefresh'
/usr/bin/ld: /tmp/go-link-161865402/000014.o: in function `_cgo_cdf43a2df1d5_Cfunc_igSetBgColor':
/tmp/go-build/sdl_backend.cgo2.c:249:(.text+0x1ce): undefined reference to `igSetBgColor'
/usr/bin/ld: /tmp/go-link-161865402/000014.o: in function `_cgo_cdf43a2df1d5_Cfunc_igSetTargetFPS':
/tmp/go-build/sdl_backend.cgo2.c:262:(.text+0x1e7): undefined reference to `igSetTargetFPS'
collect2: error: ld returned 1 exit status
```

**Comment by damntourists** [comment]
I started a fork at https://github.com/damntourists/cimgui-go-lite which now compiles with the ebiten golang library. all contents of the third party directory and the *_backend.* files have been removed.

**Comment by gucio321** [comment]
Sounds great. I would be awesome to have imgui working on ebitengine. In case of your problem: you need to disable glfw. In glfw_backend.go on the top of the file is a special go: build directive that allows to exclude this time of a special go tag is set.

**Comment by gucio321** [comment]
Ok, I'll check

**Comment by gucio321** [comment]
ok, I see, In my opinion texture.go should not depend on any backend - its completely unrelated wrapper.

**Comment by gucio321** [comment]
Thank you for this! I've opened a PR - should be merged soon.

**Comment by damntourists** [comment]
Nice! Thanks for applying that fix! Now I can point back to this repository :D

**Comment by gucio321** [comment]
awesome! let me know if you have some more questions.
Also, we'd be interessted in putting your solution in this repository/link in our readme as I hope that ebiten compilation will be much faster than our current backend. :smile:

**Comment by gucio321** [comment]
fixed in #301
